### PR TITLE
cpk for datalake

### PR DESCRIPTION
**Pull request by microzchang**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 1]
```diff
@@ -21,16 +21,44 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     * @brief SHA256 hash of the AES256 encryption key.
+     */
+    std::vector<uint8_t> KeyHash;
+
+    /**
+     * @brief The algorithm for Azure Blob Storage to encrypt with.
+     */
+    Models::EncryptionAlgorithmType Algorithm;
+  };
+
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:51]: Why not just typedef the one in blob sdk?

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp [hunk 2]
```diff
@@ -353,13 +360,27 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+      blobEncryptionKey = _detail::GetBlobEncryptionKey(m_customerProvidedKey.Value());
+    }
+    auto renamedBlobClient = Blobs::BlobClient(
+        _detail::GetBlobUrlFromUrl(destinationDfsUrl), m_pipeline, blobEncryptionKey);
     auto renamedFileClient = DataLakeFileClient(
-        std::move(destinationDfsUrl), std::move(renamedBlobClient), m_pipeline);
+        std::move(destinationDfsUrl),
+        std::move(renamedBlobClient),
+        m_pipeline,
+        m_customerProvidedKey);
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp:383]: Is the CPK preserved for the destination after renaming? Can you confirm that?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 3]
```diff
@@ -21,16 +21,44 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
   /**
    * @brief Client options used to initialize all DataLake clients.
    */
   struct DataLakeClientOptions final : Azure::Core::_internal::ClientOptions
   {
+    /**
+     * @brief Holds the customer provided key used when making requests.
+     */
+    Azure::Nullable<EncryptionKey> CustomerProvidedKey;
+
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:61]: Let's move newly added field to the end of this struct.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/rest_client.hpp [hunk 4]
```diff
@@ -374,6 +374,7 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
         Nullable<DateTime> SourceIfUnmodifiedSince;
         Nullable<std::string> EncryptionKey;
         Nullable<std::vector<uint8_t>> EncryptionKeySha256;
+        Nullable<std::string> EncryptionAlgorithm;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/rest_client.hpp:377]: We need to test only these 3 operations support CPK, and other operations don't support CPK.

File: sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp [hunk 5]
```diff
@@ -397,6 +398,133 @@ namespace Azure { namespace Storage { namespace Test {
     }
   }
 
+  TEST_F(DataLakeFileSystemClientTest, CustomerProvidedKey_LIVEONLY_)
+  {
+    // will skip test under some cased where test can't run (usually LIVE only tests)
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp:403]: Why is this live-only?

File: sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp [hunk 6]
```diff
@@ -397,6 +398,133 @@ namespace Azure { namespace Storage { namespace Test {
...
+    CHECK_SKIP_TEST();
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp:404]: Do we need this statement here even if it's live-only?

File: sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp [hunk 7]
```diff
@@ -397,6 +398,133 @@ namespace Azure { namespace Storage { namespace Test {
...
+    Azure::Core::IO::MemoryBodyStream bodyStream(buffer.data(), buffer.size());
+
+    auto customerProvidedKey = getRandomCustomerProvidedKey();
+    Files::DataLake::DataLakeClientOptions options;
+    options.CustomerProvidedKey = customerProvidedKey;
+    auto fileServiceClient
+        = Azure::Storage::Files::DataLake::DataLakeServiceClient::CreateFromConnectionString(
+            AdlsGen2ConnectionString(), options);
+    auto fileSystemClient = fileServiceClient.GetFileSystemClient(m_fileSystemName);
+
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp:429]: Need to verify CPK option can be inherited when getting filesystem client from service client.

Also need to verify CPK option is correctly set when constructing pathclient, fileclient, directoryclient from construcots (factory methods).

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp (hunk 2 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp:383]: Yes , the CPK preserved, I have tested for this scenario.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 1 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:51]: That's a good idea. I will change it.

File: sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp (hunk 5 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_system_client_test.cpp]: I recorded and find that the google test can't record all the request. Just create/delete filesystem requests were recorded. So playback failed.

**Review by Jinming-Hu — APPROVED** [review]

**Pull request merged**